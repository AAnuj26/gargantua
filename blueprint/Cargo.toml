[package]
name = "nova"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true
serde.workspace = true
serde_json = "1.0.128"
