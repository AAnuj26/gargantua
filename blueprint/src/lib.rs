mod blueprint;
mod parse;
pub use blueprint::*;

pub use url;