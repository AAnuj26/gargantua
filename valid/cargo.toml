[package]
name = "valid"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true

[dev-dependencies]
pretty_assertions.workspace = true
