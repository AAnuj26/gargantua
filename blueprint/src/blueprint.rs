use std::collections::{BTreeMap, BTreeSet};

use async_graphql_parser::types::ServiceDocument;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use valid::Valid;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Blueprint {
    pub definitions: Vec<Definition>,
    pub schema: SchemaDefinition,
    pub directives: Vec<DirectiveDefinition>,
    pub graphs: Vec<JoinGraph>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct GraphId(pub String);

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct JoinGraph {
    pub name: GraphId,
    pub url: url::Url,
}

impl Blueprint {
    pub fn parse(doc: ServiceDocument) -> Valid<Blueprint, String> {
        super::parse::parse(doc)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Definition {
    Interface(InterfaceTypeDefinition),
    Object(ObjectTypeDefinition),
    InputObject(InputObjectTypeDefinition),
    Scalar(ScalarTypeDefinition),
    Enum(EnumTypeDefinition),
    Union(UnionTypeDefinition),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InterfaceTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub implements: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<InputFieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub enum_values: Vec<EnumValueDefinition>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumValueDefinition {
    pub description: Option<String>,
    pub name: String,
    pub directives: Vec<Directive>,
    pub join_enum: Vec<GraphId>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SchemaDefinition {
    pub query: Option<String>,
    pub mutation: Option<String>,
    pub subscription: Option<String>,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputFieldDefinition {
    pub name: String,
    pub of_type: Type,
    pub default_value: Option<Value>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub name: String,
    pub args: Vec<InputFieldDefinition>,
    pub of_type: Type,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Directive {
    pub name: String,
    pub arguments: BTreeMap<String, Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DirectiveDefinition {
    pub name: String,
    pub description: Option<String>,
    pub arguments: Vec<InputFieldDefinition>,
    pub repeatable: bool,
    pub locations: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ScalarTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UnionTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub types: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_union: Vec<JoinUnion>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinType {
    pub graph: GraphId,
    pub key: Option<String>,
    pub extension: bool,
    pub resolvable: bool,
    pub is_interface_object: bool,
}

impl JoinType {
    pub fn new(graph: String) -> Self {
        Self {
            graph: GraphId(graph),
            key: None,
            extension: false,
            resolvable: true,
            is_interface_object: false,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinField {
    pub graph: Option<GraphId>,
    pub requires: Option<String>,
    pub provides: Option<String>,
    pub r#type: Option<String>,
    pub external: Option<bool>,
    pub r#override: Option<String>,
    pub used_overridden: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinImplements {
    pub graph: GraphId,
    pub interface: String
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinUnion {
    pub graph: GraphId,
    pub member: String
}

/// Type to represent GraphQL type usage with modifiers
/// [spec](https://spec.graphql.org/October2021/#sec-Wrapping-Types)
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Type {
    Named {
        /// Name of the type
        name: String,
        /// Flag to indicate the type is required.
        required: bool,
    },
    List {
        /// Type is a list
        of_type: Box<Type>,
        /// Flag to indicate the type is required.
        non_null: bool,
    },
}
