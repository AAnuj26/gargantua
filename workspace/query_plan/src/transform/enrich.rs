use async_graphql_value::Value;
use blueprint::{Index, QueryField};
use valid::{Transform, Valid, Validator};

use crate::SelectionSet;

struct Enrich(Index);

impl Enrich {
    pub fn new(index: Index) -> Self {
        Self(index)
    }

    fn enrich_information(
        &self,
        selection: SelectionSet<Value>,
        container_type: &str,
    ) -> Valid<SelectionSet<Value>, String> {
        // this field belongs to container_type, so we if want to get this field
        let type_def = match self.0.get_object_type_definition(container_type) {
            Some(type_def) => type_def,
            None => {
                return Valid::fail(format!(
                    "type definition not found for type '{}' ",
                    container_type
                ))
            }
        };

        let mut enriched_selection_set: SelectionSet<Value> = SelectionSet::default();

        for field in selection.into_vec().into_iter() {
            let mut enriched_field = field.clone();
            let field_def = match self.0.get_field(container_type, &field.name) {
                Some(QueryField::Field((def, _))) => def,
                _ => {
                    return Valid::fail(format!(
                        "field definition not found for field '{}' in type '{}' ",
                        field.name, container_type
                    ))
                }
            };

            let enriched_field = if field_def.join_fields.is_empty() {
                // if field doesn't have @join__field directive, then
                // we need to figure out from where this field can be queried

                // 1. this field can be queried form the @join__type -> wherein the key is same as this field.
                // 2. this field can be queried from the @join__type directive's graph where key is none.
                enriched_field
                    .graph
                    .extend(type_def.join_types.iter().filter_map(|jt| {
                        if jt.key.is_none() || jt.key.as_ref().map_or(false, |k| k == &field.name) {
                            Some(jt.graph.clone())
                        } else {
                            None
                        }
                    }));

                enriched_field
            } else {
                enriched_field.join_field(field_def.join_fields.clone())
            };

            let type_name = field_def.of_type.as_type_str();
            if !field.selections.is_empty() {
                self.enrich_information(field.selections, &type_name)
                    .and_then(|enriched_nested_selected_set| {
                        let enriched_field =
                            enriched_field.selections(enriched_nested_selected_set);
                        enriched_selection_set.push(enriched_field);

                        Valid::succeed(())
                    });
            } else {
                enriched_selection_set.push(enriched_field);
            }
        }

        Valid::succeed(enriched_selection_set)
    }
}

impl Transform for Enrich {
    type Value = SelectionSet<Value>;
    type Error = String;

    fn transform(&self, value: Self::Value) -> valid::Valid<Self::Value, Self::Error> {
        // TODO: fix the container type to be dynamic
        self.enrich_information(value, "Query")
    }
}

#[cfg(test)]
mod test {
    use blueprint::Blueprint;

    use super::*;

    fn setup(graphql: &str) -> Index {
        let document = async_graphql_parser::parse_schema(graphql).unwrap();
        let index = Blueprint::parse_doc(document).to_index();
        index
    }

    #[test]
    fn test_enricher_supergraph_1() {
        let query = "query { topProducts { productName: name reviews { body } reviews { id } } }";
        let index = setup(include_str!("../../../blueprint/src/fixtures/router.graphql"));
        let doc = async_graphql_parser::parse_query(query).unwrap();

        // pick the very first operation.
        let op = doc
            .operations
            .iter()
            .next()
            .unwrap()
            .1
            .node
            .selection_set
            .node
            .clone();

        let selection_set: SelectionSet<async_graphql_value::Value> =
            super::SelectionSet::from(&op);

        let enriched_selection_set = Enrich::new(index)
            .transform(selection_set)
            .to_result()
            .unwrap();

        insta::assert_debug_snapshot!(enriched_selection_set)
    }
}
