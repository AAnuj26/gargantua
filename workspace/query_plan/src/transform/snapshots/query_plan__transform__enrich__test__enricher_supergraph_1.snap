---
source: workspace/query_plan/src/transform/enrich.rs
expression: enriched_selection_set
---
SelectionSet(
    [
        Field {
            name: "topProducts",
            selections: SelectionSet(
                [
                    Field {
                        name: "name",
                        selections: SelectionSet(
                            [],
                        ),
                        arguments: [],
                        directives: [],
                        is_hidden: false,
                        graph: [],
                        join_field: [
                            JoinField {
                                graph: Some(
                                    Graph(
                                        "PRODUCTS",
                                    ),
                                ),
                                requires: None,
                                provides: None,
                                type: None,
                                external: None,
                                override: None,
                                used_overridden: None,
                            },
                        ],
                    },
                    Field {
                        name: "reviews",
                        selections: SelectionSet(
                            [
                                Field {
                                    name: "body",
                                    selections: SelectionSet(
                                        [],
                                    ),
                                    arguments: [],
                                    directives: [],
                                    is_hidden: false,
                                    graph: [
                                        Graph(
                                            "REVIEWS",
                                        ),
                                    ],
                                    join_field: [],
                                },
                            ],
                        ),
                        arguments: [],
                        directives: [],
                        is_hidden: false,
                        graph: [],
                        join_field: [
                            JoinField {
                                graph: Some(
                                    Graph(
                                        "REVIEWS",
                                    ),
                                ),
                                requires: None,
                                provides: None,
                                type: None,
                                external: None,
                                override: None,
                                used_overridden: None,
                            },
                        ],
                    },
                    Field {
                        name: "reviews",
                        selections: SelectionSet(
                            [
                                Field {
                                    name: "id",
                                    selections: SelectionSet(
                                        [],
                                    ),
                                    arguments: [],
                                    directives: [],
                                    is_hidden: false,
                                    graph: [
                                        Graph(
                                            "REVIEWS",
                                        ),
                                    ],
                                    join_field: [],
                                },
                            ],
                        ),
                        arguments: [],
                        directives: [],
                        is_hidden: false,
                        graph: [],
                        join_field: [
                            JoinField {
                                graph: Some(
                                    Graph(
                                        "REVIEWS",
                                    ),
                                ),
                                requires: None,
                                provides: None,
                                type: None,
                                external: None,
                                override: None,
                                used_overridden: None,
                            },
                        ],
                    },
                ],
            ),
            arguments: [],
            directives: [],
            is_hidden: false,
            graph: [],
            join_field: [
                JoinField {
                    graph: Some(
                        Graph(
                            "PRODUCTS",
                        ),
                    ),
                    requires: None,
                    provides: None,
                    type: None,
                    external: None,
                    override: None,
                    used_overridden: None,
                },
            ],
        },
    ],
)
